### ISSUE TITLE
Add Command-Line Interface (CLI) for BardGPT Training & Inference

### ISSUE THREAD

**Issue by Himanshu7921** [Himanshu7921/BardGPT#7]
## 1. Issue Type

* [ ] Bug Report
* [x] Feature Request
* [x] Documentation Improvement
* [ ] Training / Sampling Issue
* [ ] Performance / Optimization Request
* [ ] Other

---

## 2. Description

BardGPT currently requires users to **edit `train.py` manually** to change:

* dataset path
* steps
* batch size
* model size
* learning rate
* sampling options
* checkpoint paths
* warmup settings

This makes experimentation and reproducibility more difficult.

Adding a **Command-Line Interface (CLI)** will allow users to configure training and sampling directly from the terminal.

This will significantly improve usability, contributor onboarding, and workflow automation.

---

## 3. Steps to Reproduce (for bugs)

*Not a bug — feature enhancement request.*

---

## 4. Environment Information

* OS: Any
* Python version: 3.10+
* PyTorch version: Any
* CUDA version: Optional
* GPU model: Optional
* BardGPT commit hash or version: Latest
* Installed dependencies: N/A

---

### Example of the CLI that should be supported:

```bash
bardgpt train \
    --dataset ./data/tiny_shakespeare.txt \
    --steps 200000 \
    --batch-size 64 \
    --lr 3e-4 \
    --seq-length 128 \
    --model-dim 384 \
    --n-layers 6 \
    --n-heads 6 \
    --dropout 0.2 \
    --save-every 5000
```

### Example for sampling:

```bash
bardgpt generate \
    --checkpoint checkpoints/best_val_0072000.pt \
    --prompt "ROMEO:" \
    --temperature 0.7 \
    --top-k 40 \
    --tokens 300
```

### Why this is important:

* Makes BardGPT easier to run on servers, Colab, scripts
* Helps contributors test various configurations without editing code
* Provides more professional developer experience
* Enables reproducible experiments
* Makes BardGPT behave like real ML frameworks

---

### **Contributor Tasks**

* [ ] Add CLI using `argparse` or `typer` (recommended)
* [ ] Add `bardgpt train` command
* [ ] Add `bardgpt generate` command
* [ ] Add config loading from CLI flags
* [ ] Add help screens (`bardgpt train --help`)
* [ ] Add validation for parameters
* [ ] Add entry point in `pyproject.toml` or `setup.py`
* [ ] Update README with CLI usage
* [ ] Ensure existing training pipeline integrates cleanly

---

### **Suggested Folder Structure**

```
bardgpt/
 ├── cli/
 │    ├── train_cli.py
 │    ├── generate_cli.py
 │    ├── __init__.py
 └── __main__.py
```

---

### Additional Optional Features

* YAML config loading:
  `bardgpt train --config config/base.yaml`
* Automatic checkpoint resume
* Ability to override config values
* Rich-style colored progress output

---

Thank you for taking the time to report this issue.
Your contribution helps improve BardGPT for everyone.
